src/System.Net.Http/src/System/Net/Http/Unix/CurlHandler.MultiAgent.cs

```diff
@@ -0,0 +1,658 @@
+                    }
+                }
+
+                // Returing a value other than size fails the callback and forces
+                // request completion with an error
+                return size - 1;
+            }
+
+            private unsafe static size_t CurlReceiveBodyCallback(
+                byte* buffer, size_t size, size_t nitems, IntPtr context)
+            {
+                size *= nitems;
+
+                EasyRequest easy;
+                if (TryGetEasyRequestFromContext(context, out easy))
+                {
+                    try
+                    {
+                        if (!(easy.Task.IsCanceled || easy.Task.IsFaulted))
+                        {
+                            // Complete the task if it hasn't already been.  This will make the
+                            // stream available to consumers.  A previous write callback
+                            // may have already completed the task to publish the response.
+                            easy.EnsureResponseMessagePublished();
+
+                            // Wait for a reader
+                            // TODO: The below call blocks till all the data has been read since
+                            //       response body is not suppored to be buffered in memory.
+                            //       Figure out some way to work around this
+                            if (size != 0)
+                            {
+                                easy.ResponseMessage.ContentStream.WaitAndSignalReaders(buffer, (long)size);
+                            }
+
+                            return size;
+                        }
+                    }
+                    catch (Exception ex)
+                    {
+                        easy.FailRequest(ex); // cleanup will be handled by main processing loop
+                    }
+                }
+
+                // Returing a value other than size fails the callback and forces
+                // request completion with an error.
+                return (size > 0) ? size - 1 : 1;
+            }
+
+            private static size_t CurlSendCallback(IntPtr buffer, size_t size, size_t nitems, IntPtr context)
+            {
+                size *= nitems;
+                if (size == 0)
+                {
+                    return 0;
+                }
+
+                EasyRequest easy;
+                if (TryGetEasyRequestFromContext(context, out easy))
+                {
+                    Debug.Assert(easy.RequestContentStream != null, "We should only be in the send callback if we have a request content stream");
+                    Debug.Assert(easy._associatedMultiAgent != null, "The request should be associated with a multi agent.");
+
+                    // Transfer data from the request's content stream to libcurl
+                    try
+                    {
+                        byte[] arr = easy._associatedMultiAgent._transferBuffer;
```

Comment:
Everything on the multi handle is processed by the single thread.  There shouldn't be any problem using a single buffer, and this saves us from having to allocate a 16KB buffer per request, instead doing it per agent and reusing that buffer across however many requests it processes.  Is there a particular case you're concerned about?